### PR TITLE
KAFKA-14462; [16/N] Add CoordinatorLoader and CoordinatorPlayback interfaces

### PR DESCRIPTION
**Pull request by dajac**
This patch adds the CoordinatorLoader and the CoordinatorPlayback interfaces. The former is used to load (or rebuild) the coordinator state by replaying all the records stored in the partition. The later is the interface that must be implemented by coordinator to support replaying records.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jolshan** [comment]
Let's add a brief description to explain the rationale for these interfaces to the description

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.runtime;
+
+/**
+ * The Replayable interface.
+ */
```
Comment (jolshan) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java]: nit -- could we add a bit more detail here? Maybe something about replaying records to generate state etc

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public interface Replayable<U> {
```
Comment (CalvinLiu7947) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java]: Do we want to name it more specifically to the coordinators?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java (hunk 1 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java]: Done.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java (hunk 2 above)
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/Replayable.java]: I have renamed it to `CoordinatorPlayback`. Does it sound good to you?

**Review by jolshan — APPROVED** [review]
Thanks!

**Pull request merged**